test/source-updater.test.js

```diff
@@ -1,232 +1,769 @@
+
+  assert.ok(abortedAudio, 'aborted audio');
+  assert.ok(abortedVideo, 'aborted video');
+  assert.notOk(this.sourceUpdater.audioBuffer, 'removed audioBuffer reference');
+  assert.notOk(this.sourceUpdater.videoBuffer, 'removed videoBuffer reference');
+});
+
+QUnit.test('no error passed by default in done callback', function(assert) {
+  const done = assert.async();
+
+  this.sourceUpdater.createSourceBuffers({
+    audio: 'mp4a.40.2'
+  });
+
+  this.sourceUpdater.appendBuffer('audio', mp4Audio(), (error) => {
+    assert.notOk(error, 'no error');
+    done();
+  });
+});
+
+// Firefox allows appending video to an audio buffer, for some reason.
+// TODO find an alternative for triggering an error
+QUnit.skip('audio source buffer error passed in done callback', function(assert) {
+  const done = assert.async();
+
+  this.sourceUpdater.createSourceBuffers({
+    audio: 'mp4a.40.2'
+  });
+
+  // errors when appending video to an audio buffer
+  this.sourceUpdater.appendBuffer('audio', mp4Video(), (error) => {
+    assert.ok(error, 'error passed back');
+    done();
+  });
+});
+
+// Firefox allows appending audio to a video buffer, for some reason.
+// TODO find an alternative for triggering an error
+QUnit.skip('video source buffer error passed in done callback', function(assert) {
+  const done = assert.async();
+
+  this.sourceUpdater.createSourceBuffers({
+    video: 'avc1.4D001E'
+  });
+
+  // errors when appending audio to a video buffer
+  this.sourceUpdater.appendBuffer('video', mp4Audio(), (error) => {
+    assert.ok(error, 'error passed back');
+    done();
+  });
+});
+
+QUnit.test('setDuration processes immediately if not waiting on source buffers',
+function(assert) {
+  this.sourceUpdater.createSourceBuffers({
+    audio: 'mp4a.40.2',
+    video: 'avc1.4D001E'
+  });
+
+  assert.ok(Number.isNaN(this.mediaSource.duration), 'duration set to NaN at start');
+  this.sourceUpdater.setDuration(11);
+  assert.equal(this.mediaSource.duration, 11, 'set duration on media source');
+});
+
+QUnit.test('setDuration waits for audio buffer to finish updating', function(assert) {
+  const done = assert.async();
+
+  assert.expect(5);
+
+  this.sourceUpdater.createSourceBuffers({
+    audio: 'mp4a.40.2',
+    video: 'avc1.4D001E'
+  });
 
-  updater.abort();
-  this.mediaSource.trigger('sourceopen');
-  assert.equal(updater.callbacks_.length,
-               0,
-               'abort not queued before source buffers are appended to');
+  assert.notOk(this.sourceUpdater.updating(), 'not updating by default');
 
-  updater.appendBuffer(new Uint8Array([0]));
+  const checkDuration = () => {
+    // duration is set to infinity if content is appended before an explicit duration is
+    // set https://w3c.github.io/media-source/#sourcebuffer-init-segment-received
+    assert.equal(this.mediaSource.duration, Infinity, 'duration not set on media source');
+  };
 
-  updater.abort();
+  this.sourceUpdater.appendBuffer('audio', mp4Audio(), checkDuration);
```

Comment:
might be a bit easier to follow if an anonymous function was used here instead

Reply:
I just didn't want to copy the comment and code inside of the function for both video and audio, but am open to changing if you think the copying makes it easier to read.

Reply:
I think it'll make it easier to read

Reply:
I was looking at a different test where we reused it. Definitely makes sense as anonymous here 👍

Reply:
(Possibly there as well.)